### ISSUE TITLE
Significant memory spike in plugin with each()

### ISSUE THREAD

**Issue by cjheppell-ravio** [graphile/crystal#2038]
### Summary

Javascript heap memory usage jumps significantly when using a plugin with `loadOne`, . This occurs for relatively "small" rows. ~100 rows will cause a jump from ~40mb on startup to ~250mb. Larger numbers cause higher memory usage.

For our usage of Postgraphile V5 in production, this is causing our server to crash due to out of memory errors.

### Steps to reproduce

See this repository for a full reproduction:

https://github.com/cjheppell/grafast-memory-spike-repro/tree/main

### Expected results

Memory usage does not jump significantly.

### Actual results

Memory usage jumps significantly.

### Additional context

Postgres 15, Node v20, MacOS M1 32GB, latest Postgraphile v5.

Schema, data and code is available in the reproduction repo above.

### Possible Solution

Unknown. I'm too unfamiliar with the internals of Grafast/Postgraphile to suggest how this might be resolved.

**Comment by cjheppell-ravio** [comment]
Update: This appears to not be related to `applyTransform` _or_ `loadOne` at all. There's a simpler repro here with just `each`: https://github.com/scottravio/grafast-memory-spike-repro

**Comment by benjie** [comment]
This is due to the `each()` creating a subroutine (since it's not used as the return result of a plan resolver, but instead as input to another step), and subroutines are currently not hoisted (see #2041); as such we're having to run the `each` transform with a batch size of 1000 (because you're fetching 1000 users, and this is a field in the `User` type), and for each of those 1,000 user we're then mapping over the 1000 languages ultimately resulting in 1000*1000 = 1,000,000 values.

Since your step here is not dependent on `$user`, theoretically we should be able to hoist it (see #2041) but we currently do not. Instead, try and ensure that the steps that you use can all be hoisted (see the plan diagram to see which bucket the steps exist in); for example:

```ts
    plans: {
      User: {
        totalLanguages() {
          const $allLangugages = withPgClient(
            executor,
            constant(null),
            async (pgClient) => {
              const { rows } = await pgClient.query({
                text: `select language from languages`,
              });
              return rows;
            },
          );
          $allLangugages.hasSideEffects = false;
          return lambda(
            $allLangugages,
            (allLangugages) => allLangugages.length,
          );
        },
      },
    },
```

However, it would be better to directly address your need, which in this case seems to be counting the number of languages on a per-user basis, so perhaps something like:

```ts
    plans: {
      User: {
        totalLanguages($user) {
          const $agg = user_config.find({ user_id: $user.get("id") }).clone("aggregate");
          return $agg.single().select(sql`count(*)`, TYPES.bigint);
        },
      },
    },
```

Note that `.clone` is currently marked as an internal API, but that is resolved in #2039.

**Comment by benjie** [comment]
Closing this issue in favour of the related issues #2039, #2040 and #2041.
